node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+			per_candidate
+				.sent_messages
+				.entry(peer)
+				.or_default()
+				.insert(message_id);
+		}
+
+		let encoded = message.encode();
+		per_candidate
+			.message_vault
+			.insert(message.erasure_chunk.index, message);
+
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::SendMessage(
+				peers.clone(),
+				AvailabilityDistributionSubsystem::PROTOCOL_ID,
+				encoded,
+			),
+		))
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	}
+
+	Ok(())
+}
+
+// Send the difference between two views which were not sent
+// to that particular peer.
+async fn handle_peer_view_change<Context>(
+	ctx: &mut Context,
+	state: &mut ProtocolState,
+	origin: PeerId,
+	view: View,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let current = state.peer_views.entry(origin.clone()).or_default();
+
+	let delta_vec: Vec<H256> = (*current).difference(&view).cloned().collect();
+
+	*current = view;
+
+	// only contains the intersection of what we are interested and
+	// the union of all relay parent's candidates.
+	let delta_candidates = state.cached_live_candidates_unioned(delta_vec.iter());
+
+	// Send all messages we've seen before and the peer is now interested
+	// in to that peer.
+
+	for (candidate_hash, receipt) in delta_candidates {
+		let per_candidate = state.per_candidate.entry(candidate_hash).or_default();
+
+		let messages = if let Some(per_relay_parent) = state
+			.per_relay_parent
+			.get_mut(&receipt.descriptor().relay_parent)
+		{
+			// obtain the relevant chunk indices not sent yet
+			let messages = ((0 as ValidatorIndex)
+				..(per_relay_parent.validators.len() as ValidatorIndex))
+				.into_iter()
+				.filter(|erasure_chunk_index: &ValidatorIndex| {
+					// check if that erasure chunk was already sent before
+					if let Some(sent_set) = per_candidate.sent_messages.get(&origin) {
+						!sent_set.contains(&(candidate_hash, *erasure_chunk_index))
+					} else {
+						true
+					}
+				})
+				.filter_map(|erasure_chunk_index: ValidatorIndex| {
+					// try to pick up the message from the message vault
+					per_candidate.message_vault.get(&erasure_chunk_index)
+				})
+				.cloned()
+				.collect::<HashSet<_>>();
+			messages
+		} else {
+			continue;
+		};
+
+		send_tracked_gossip_messages_to_peer(ctx, per_candidate, origin.clone(), messages).await?;
+	}
+	Ok(())
+}
+
+/// Obtain the first key with a signing key, which must be ours. We obtain the index as `ValidatorIndex`
+/// If we cannot find a key in the validator set, which we could use.
+fn obtain_our_validator_index<Context>(
+	validators: &[ValidatorId],
+	keystore: KeyStorePtr,
+) -> Option<ValidatorIndex> {
+	let keystore = keystore.read();
+	validators.iter().enumerate().find_map(|(idx, v)| {
+		keystore
+			.key_pair::<ValidatorPair>(&v)
+			.ok()
+			.map(move |_| idx as ValidatorIndex)
+	})
+}
+
+/// Handle an incoming message from a peer.
+async fn process_incoming_peer_message<Context>(
```

Comment:
This function should also be issuing a `StoreChunk` message to the availability store